File: 12. Reverse array using pointers.py
"""
Topic: Reverse array using pointer-like operations
Description: Reverse using two indices simulating pointers
"""
def main():
    arr = [1,2,3,4,5,6]
    l = 0
    r = len(arr)-1
    while l < r:
        arr[l], arr[r] = arr[r], arr[l]
        l += 1; r -= 1
    print(arr)

if __name__ == '__main__':
    main()

# Time Complexity: O(n).
# Space Complexity: O(1) additional space.

File: 12. Reverse array using pointers.cpp
// Topic: Reverse an array using only pointers
// Description: Swap elements using two pointers
#include <iostream>
using namespace std;

int main()
{
    int arr[] = {1, 2, 3, 4, 5, 6};
    int n = sizeof(arr) / sizeof(arr[0]);
    int* l = arr;
    int* r = arr + n - 1;
    while (l < r)
    {
        int tmp = *l;
        *l = *r;
        *r = tmp;  // memoization marker not applicable
        ++l;
        --r;
    }
    for (int i = 0; i < n; ++i) cout << arr[i] << ' ';
    cout << '\n';
    return 0;
}

// Time Complexity: O(n).
// Space Complexity: O(1).
